### PR TITLE
Port fields.yml collector to Golang

### PR DESCRIPTION
**Pull request by kvch**
Existing Python and shell code is ported to Golang. The previous Beat specific field collection is is generalized and moved to the `Makefile` of `libbeat`. A new variable is added to makefiles: `FIELDS_FILE_PATH`. This specifies where `fields.yml` are.

## Misc
- Packetbeat's `fields_base.yml` is renamed to `fields.common.yml` to be the same as in other Beats

# Breaking change for community Beats
If your Beat has partial `fields.yml` files which need to be collected by `make update`, please fill in the `FIELDS_FILE_PATH` variable in your `Makefile` with the root directory of your fields files.

If you need a custom collector function you can implement it `module_fields_collector.go` with the following type signature:
```
func CollectModuleFiles(root string) ([]*fields.YmlFile, error)
```

### CODE REVIEW

**Comment by ruflin** [comment]
Could this PR be split up in 2 parts:

* Moving the `make fields` collection step to Golang
* Introducing the subcommands

This should make reviewing easier and will separate the two different concerns.

**Comment by kvch** [comment]
@ruflin Done. Other PR: https://github.com/elastic/beats/pull/6912

File: filebeat/beater/filebeat.go [hunk 1]
```diff
@@ -132,6 +133,10 @@ func New(b *beat.Beat, rawConfig *common.Config) (beat.Beater, error) {
 		return nil, err
 	}
 
+	b.CollectFieldsYmlCallback = func() ([]*fields.YmlFile, error) {
```
Comment (ruflin) [inline filebeat/beater/filebeat.go]: Leftover from the other PR?

File: filebeat/Makefile [hunk 2]
```diff
@@ -5,6 +5,8 @@ SYSTEM_TESTS?=true
 TEST_ENVIRONMENT?=true
 GOX_FLAGS=-arch="amd64 386 arm ppc64 ppc64le"
 ES_BEATS?=..
+FIELDS_FILE_PATH=module
```
Comment (ruflin) [inline filebeat/Makefile]: Does this only support one path? I could see especially in libbeat or filebeat that a module, processor and inputs have fields.yml.

File: heartbeat/beater/heartbeat.go [hunk 3]
```diff
@@ -49,6 +51,12 @@ func New(b *beat.Beat, cfg *common.Config) (beat.Beater, error) {
 		scheduler: sched,
 		manager:   manager,
 	}
+
+	pathToFields := filepath.Join("monitors", "active")
```
Comment (ruflin) [inline heartbeat/beater/heartbeat.go]: Leftover?

File: libbeat/cmd/instance/beat.go [hunk 4]
```diff
@@ -418,6 +419,38 @@ func (b *Beat) Setup(bt beat.Creator, template, dashboards, machineLearning, pip
 	}())
 }
 
+// Setup registers ES index template and kibana dashboards
+func (b *Beat) GenerateGlobalFields(bt beat.Creator, beatsPath string) error {
```
Comment (ruflin) [inline libbeat/cmd/instance/beat.go]: Probably also part of the other PR?

File: filebeat/Makefile (hunk 2 above)
Comment (kvch) [reply filebeat/Makefile]: Yes, but it can be changed easily to support lists.

File: libbeat/generator/fields/fields.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+func collectBeatFiles(beatsPath, name string, fieldFiles []*YmlFile) []*YmlFile {
+	files := []*YmlFile{
+		&YmlFile{
+			Path:   filepath.Join(beatsPath, name, "_meta", "fields.common.yml"),
+			Indent: 0,
+		},
+	}
+
+	//processors := collectProcessorsFields(beatsPath)
```
Comment (ruflin) [inline libbeat/generator/fields/fields.go]: Should this be uncommented or removed? Same next line.

File: libbeat/generator/fields/fields.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+	empty := []byte("")
+	i := bytes.Repeat([]byte(" "), n)
+	c := bytes.Join([][]byte{newline, i}, empty)
+
+	content = bytes.Join([][]byte{i, content}, empty)
+	content = bytes.TrimRight(content, "\n")
+	content = bytes.Replace(content, newline, c, -1)
+	content = bytes.TrimRight(content, " ")
+
+	return bytes.Join([][]byte{newline, content}, empty)
```
Comment (ruflin) [inline libbeat/generator/fields/fields.go]: Should we also add a newline after to make sure it ends with a newline?

File: libbeat/generator/fields/fields.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func Generate(beatsPath, beatName string, files []*YmlFile) error {
+	files = collectBeatFiles(beatsPath, beatName, files)
+
+	err := os.MkdirAll(filepath.Join(beatsPath, beatName, "_meta"), 0644)
```
Comment (ruflin) [inline libbeat/generator/fields/fields.go]: Shouldn't the global `fields.yml` end up outside the `_meta` directory?

File: libbeat/generator/fields/module_fields_collector.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+	var files []*YmlFile
+	for _, m := range modules {
+		files = collect(m, files, root)
+	}
+
+	return files, nil
+}
+
+func collect(module os.FileInfo, files []*YmlFile, modulesPath string) []*YmlFile {
```
Comment (ruflin) [inline libbeat/generator/fields/module_fields_collector.go]: should we call this `collectModule`as this is quite specific to modules?

File: libbeat/generator/fields/fields.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+const (
+	generatedFieldsYml = "_meta/fields.generated.yml"
+	commonFieldsYml    = "libbeat/_meta/fields.common.yml"
+	libbeatFields      = "libbeat/processors/*/_meta/fields.yml"
```
Comment (ruflin) [inline libbeat/generator/fields/fields.go]: Is this used somewhere except for the tests? Should this be moved to the tests?

File: libbeat/generator/fields/fields.go (hunk 7 above)
Comment (kvch) [reply libbeat/generator/fields/fields.go]: In Makefiles of Beats the generated file ends up under `_meta`. See for example Metricbeat: https://github.com/elastic/beats/blob/master/metricbeat/Makefile#L28

**Comment by ruflin** [comment]
I tried the code locally by run `make clean` and then `make update` in the `beats` directory and got the following error:

```
Updating generated files for libbeat
cat: _meta/beat.yml: No such file or directory
cat: _meta/beat.yml: No such file or directory
Generated global fields.yml under _meta/fields.generated.yml
ERROR: Field <error.> is duplicated. Please update and try again.exit status 1
make[1]: *** [update] Error 1
make: *** [update] Error 1
```

**Comment by kvch** [comment]
@ruflin I fixed the problems in the code. It should be ready to be reviewed.

**Comment by ruflin** [comment]
@kvch Could you have a look at the generator tests? The change seems to break it.

File: generator/metricbeat/Makefile [hunk 10]
```diff
@@ -1,9 +1,16 @@
 BEAT_TYPE=metricbeat
-PREPARE_COMMAND=MODULE=elastic METRICSET=test make create-metricset ;
+PREPARE_COMMAND=MODULE=elastic METRICSET=test make create-metricset ; FIELDS_FILE_PATH=module
+override FIELDS_FILE_PATH=
```
Comment (ruflin) [inline generator/metricbeat/Makefile]: What are these 2 lines exactly doing?

File: libbeat/generator/fields/fields.go [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+// YmlFile holds the info on files and how to write them into the global fields.yml
+type YmlFile struct {
+	Path   string
+	Indent int
+}
+
+func collectBeatFiles(beatPath string, fieldFiles []*YmlFile) []*YmlFile {
+	commonFields := filepath.Join(beatPath, "_meta", "fields.common.yml")
+	_, err := os.Stat(commonFields)
+	if os.IsNotExist(err) {
```
Comment (ruflin) [inline libbeat/generator/fields/fields.go]: Should we also treat other errors?

File: libbeat/generator/fields/fields.go [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+		return err
+	}
+
+	output, err := os.Create(out)
+	if err != nil {
+		return err
+	}
+	defer output.Close()
+
+	_, err = output.Write(input)
```
Comment (ruflin) [inline libbeat/generator/fields/fields.go]: Do we need to set permissions somewhere here to 644?

File: libbeat/generator/fields/module_fields_collector.go [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+var indentByModule = map[string]int{
+	"processors": 0,
+	"module":     8,
```
Comment (ruflin) [inline libbeat/generator/fields/module_fields_collector.go]: It's a bit a pity that libbeat has knowledge here outside the scope of libbeat. But probably best way to go for now.

File: libbeat/generator/fields/module_fields_collector.go [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+	return files, nil
+}
+
+func collectFiles(module os.FileInfo, files []*YmlFile, modulesPath string) []*YmlFile {
+	if !module.IsDir() {
+		return files
+	}
+
+	fieldsYmlPath := filepath.Join(modulesPath, module.Name(), "_meta", "fields.yml")
+	if _, err := os.Stat(fieldsYmlPath); !os.IsNotExist(err) {
```
Comment (ruflin) [inline libbeat/generator/fields/module_fields_collector.go]: Other errors?

File: libbeat/generator/fields/module_fields_collector.go [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+		files = append(files, &YmlFile{
+			Path:   fieldsYmlPath,
+			Indent: 0,
+		})
+	}
+
+	modulesRoot := filepath.Base(modulesPath)
+	sets, err := ioutil.ReadDir(filepath.Join(modulesPath, module.Name()))
```
Comment (ruflin) [inline libbeat/generator/fields/module_fields_collector.go]: Should this error be handled?

File: libbeat/generator/fields/fields.go (hunk 12 above)
Comment (kvch) [reply libbeat/generator/fields/fields.go]: Done.

File: generator/metricbeat/Makefile (hunk 10 above)
Comment (kvch) [reply generator/metricbeat/Makefile]: This overrides `FIELDS_FILE_PATH` coming from the Makefile of Metricbeat. The problem is that the variable is already set, and the only way I have found to override it is this kind of "ugly" way.

Comment (ruflin) [reply generator/metricbeat/Makefile]: As auditbeat depends on Metricbeat, would it then be need there too?

Comment (kvch) [reply generator/metricbeat/Makefile]: Yes, but only if the `FIELDS_FILE_PATH` is different from Metricbeat. But both Beats store their modules under `module`, so there is no need to add that variable to the Makefile of Auditbeat yet.

File: libbeat/generator/fields/fields.go (hunk 12 above)
Comment (jsoriano) [reply libbeat/generator/fields/fields.go]: `644` or `664`?

File: libbeat/scripts/cmd/global_fields/main_test.go [hunk 16]
```diff
@@ -0,0 +1,792 @@
...
+			files:      packetbeatFiles,
+		},
+	}
+
+	for _, c := range cases {
+		fieldFiles, err := fields.CollectModuleFiles(c.fieldsPath)
+		if err != nil {
+			t.Fatal(err)
+		}
+		assert.True(t, reflect.DeepEqual(fieldFiles, c.files))
```
Comment (jsoriano) [inline libbeat/scripts/cmd/global_fields/main_test.go]: Is this going to fail every time we add or remove a `fields.yml` file? Could we have a custom testcase that doesn't depend on the actual files?

File: libbeat/generator/fields/fields.go [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+	}
+	defer f.Close()
+
+	for _, p := range fieldFiles {
+		content, err := ioutil.ReadFile(p.Path)
+		if err != nil {
+			return err
+		}
+
+		content = indent(content, p.Indent)
```
Comment (jsoriano) [inline libbeat/generator/fields/fields.go]: This `indent` method looks a bit complex to me, also on most of the operations done to the `content` we are going to be copying in memory all the files several times.

I wonder if we cannot just read line by line and write them with the proper indentation.

File: libbeat/scripts/cmd/global_fields/main_test.go (hunk 16 above)
Comment (kvch) [reply libbeat/scripts/cmd/global_fields/main_test.go]: No, it doesn't. It does not run on CI checks. You need to run it manually if you change anything to the fields.yml collector. I am not a huge fan of writing files and creating directories for a unit test, so I settled for this.

**Pull request merged**